README.md
# HDFSWorkspace

## Description

If you don't have that much amount of free space on your local disk but you have an Hadoop cluster not far away, you can save your R Workspace on HDFS.

This is very useful if you have shared server that share home's directories (with NFS for example) and user that use a large amount of RAM (> 10Go).

## How to use it

### Set variable environment

You have to set some environment variable for Hadoop:

* HADOOP_BIN: Where hadoop binaries are
* JAVA_HOME: Java home
* HADOOP_HOME: Hadoop home

In file /etc/R/Renviron.site:
```
HADOOP_BIN=/opt/hadoop-3.0.0-alpha4/bin
JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64/
HADOOP_HOME=/opt/hadoop-3.0.0-alpha4
```

### Automatic save and restore of workspace

To autotmaticely save and restore workspace that is on HDFS, you have to completed the .First and .Last function:

For example in file /etc/R/Rprofile.site:

```R
.First <- function() {
        if(interactive()) {
                require("HDFSWorkspace", quietly = TRUE)
                loadWorkspaceHDFS()
        }
}

.Last <- function() {
        if(interactive()) {
                require("HDFSWorkspace", quietly = TRUE)
                saveWorkspaceHDFS()
        }
}
```

# TODO

* Handle the quit(save="no"), at this time, it always save workspace. As workaround, I add saveWorkspace function to set the choosen behavior.



NEWS.md
# HDFSWorkspace 0.1.1

* Fix pv on Rstudio, to have the progession of the load/save of the workspace in R console.
* Add message when moving the RData on HDFS to his final place.

# HDFSWorkspace 0.1.0

* Save Workspace on HDFS.
* Load Workspace on HDFS.
* Disable or enable the save of the workspace when closing R (.Last).
* Automatically load or save the current Workspace if no file is giving to the saveWorkspaceHDFS/loadWorkspaceHDFS. It use the working directory to push a .RData file on HDFS like hdfs://namenode:8020/user/USER/.rdirname/WORKING_DIRECTORY/.RData.





R/main.R
# Author: Christophe NOUCHET
# Email: nouchet.christophe@gmail.com
# Date: 27/08/2017
# Infomation: Thanks to Nicolas Dupeux for the POC that inspired this package, especially for the use of pv command
# Documentation: This package is a simple way to save user workspace on HDFS instead of local disk. It's necessary if you have multiple R server than share their home on
#                nfs (or something else) and you don't have that much free space.


#' @title A little fix for file.path
#' @name fix_path
#' @description  Ok, file.path don't check if the path already have a '/' at the start or end of parameters, so you can have double "/" ...
fix_path <- function(path) {
  return(gsub("///", "/", gsub("//", "/", path)))
}


#' @title Get the R directory to use on HDFS
#' @name getRDirName
#' @description Get the path to use for creating HDFS tree
#' @return string
getRDirName <- function() {
  rdirname <- Sys.getenv("HADOOP_RDIRNAME")
  if(rdirname == ""){
    rdirname <- ".rdirname"
  }
  return(rdirname)
}

#' @title Don't save Workspace on HDFS
#' @name setSavedWorkspace
#' @description Don't save Workspace on HDFS
#' @export
setSavedWorkspace <- function(action="yes") {
  Sys.setenv(HDFSWorkspace_SAVE=action)
}

#' @title Know if we have to save Workspace
#' @name getSavedWorkspace
#' @description Know if we have to save Workspace
#' @export
getSavedWorkspace <- function() {
  return(substr(Sys.getenv("HDFSWorkspace_SAVE"), 0, 1) == "y")
}

#' @title Get the project path on HDFS
#' @name getProjectPath
#' @description Get the project path
#' @return string
getProjectPath <- function() {

  return(fix_path(file.path(getRDirName(), getwd())))
}

#' @title Load the workspace from HDFS
#' @name loadWorkspaceHDFS
#' @description Load Workspace from HDFS
#' @param filename The filename to load on HDFS. If null, filname will be getwd() + "/.RData".
#' @return boolean
#' @export
loadWorkspaceHDFS <- function(filename = NULL) {
  setSavedWorkspace(action="yes")

  # Auto name
  if(is.null(filename) && file.exists(fix_path(file.path(getwd(), ".RDataHDFS")))) {
    filename <- fix_path(file.path(getProjectPath(), ".RData"))
  }
  else {
    return(FALSE)
  }

  # Get the file on HDFS
  connection <- readHDFSFile(filename = filename)

  if(connection == FALSE){
    message("No Workspace load from HDFS")
    return(FALSE)
  }

  message("Load Workspace from HDFS: ", filename)

  # Read it and populate user workspace
  load(envir=.GlobalEnv, file = connection)
  close(connection)

  message("Load Workspace from HDFS: OK")

  return(TRUE)
}

#' @title Save Workspace on HDFS
#' @name saveWorkspaceHDFS
#' @description Save your workspace on HDFS
#' @param filename Path to the file on HDFS. If null, filname will be getwd() + "/.RData".
#' @return boolean
#' @export
saveWorkspaceHDFS <- function(filename = NULL) {

  if(!getSavedWorkspace()){
    return(FALSE)
  }

  # Auto name
  if(is.null(filename)) {
    project_path <- getProjectPath()

    #Create the directory
    hdfs_mkdir(project_path)

    # SaveWorkspace
    filename <- fix_path(file.path(project_path, ".RData"))
  }

  message("Save Workspace on HDFS: ", filename)

  # We write the file on a temp_file for more safety (if the write failed, you don't loose the previous one)
  tmp_filename = paste(filename, as.numeric(as.POSIXlt(Sys.time())), sep="-")

  # Write file on HDFS
  connection <- writeHDFSFile(tmp_filename)
  save(list = ls(envir=.GlobalEnv, all.names = TRUE), file = connection)
  close(connection)

  # Delete old file and move the new one
  message("Move the .Rdata to his final place on HDFS: [1/2]")
  suppressWarnings(system(hdfs_dfs_command(paste("-rm", filename)), ignore.stderr = TRUE, ignore.stdout = TRUE, intern = TRUE))
  message("Move the .Rdata to his final place on HDFS: [2/2]")
  suppressWarnings(system(hdfs_dfs_command(paste("-mv", tmp_filename, filename)), intern=TRUE, ignore.stderr = TRUE, ignore.stdout = TRUE))

  message("Save Workspace on HDFS: OK")

  file.create(".RDataHDFS")

  # Clean user workspace as we want that R only save it on HDFS
  rm(list=ls(envir=globalenv(), all.names = TRUE), envir = globalenv())

  return(TRUE)
}


R/hdfs.R
# Author: Christophe NOUCHET
# Email: nouchet.christophe@gmail.com
# Date: 27/08/2017
# Infomation: Thanks to Nicolas Dupeux for the POC that inspired this package, especially for the use of pv command
# Documentation: Some primitive to use HDFS. All this function use "system" to call hdfs command. To make this possible, you must set on your Renviron.site:
#                - HADOOP_HOME
#                - JAVA_HOME
#                - HADOOP_BIN

#' @title Get the absolute path to hadoop bin directory
#' @name hadoop_command
#' @description Get the bin path of Hadoop, HADOOP_HOME and HADOOP_BIN must be set in your Renviron.site
#' @param command The hadoop command to get the absolute path of
#' @return The absolute path of the Hadoop command
hadoop_command <- function(command="") {
  hadoop_bin <- Sys.getenv("HADOOP_BIN")
  if(hadoop_bin == ""){
    hadoop_bin <- "/usr/bin"
  }
  return(fix_path(file.path(hadoop_bin, command)))
}

#' @title Get the absolute path to 'hdfs' command
#' @name hdfs_command
#' @description Get the absolute path of the hdfs command
#' @param options Option to git to the hdfs command
#' @return The absolute path of the hdfs command
hdfs_command <- function(options="") {
  return(hadoop_command(paste("hdfs", options, sep=" ")))
}

#' @title Get the absolute path to 'hdfs dfs' command
#' @name hdfs_dfs_command
#' @description Get the absolute path of the hdfs command
#' @param options Option to git to the dfs command
#' @return The absolute path of the dfs command
hdfs_dfs_command <- function(options) {
  return(hdfs_command(paste("dfs", options, sep=" ")))
}

#' @title Liste file/directory on HDFS
#' @name hdfs_dfs_ls
#' @description Liste file on hadoop, get file size
#' @param options Options to git to ls
#' @return List of files/directories with their size
hdfs_ls <- function(options="") {
  # Execute hdfs command
  suppressWarnings(raw_file <- system(command = hdfs_dfs_command(paste("-ls", options, "|grep -v \"Found .*item.*\" | awk '{print $8, $5}'", sep=" ")), intern=TRUE, ignore.stderr = TRUE))

  # List of file
  liste <- list()

  # Get the file and their size
  for(i in raw_file){
    temp_list <- strsplit(i, " ")
    liste[[temp_list[[1]][1]]] <- temp_list[[1]][2]
  }
  return(liste)
}

#' @title Read a file from HDFS
#' @name readHDFSFile
#' @description Get a file on HDFS
#' @param filename The filename on HDFS to load
#' @param options File mode
#' @return A file object that you can use with load
readHDFSFile <- function(filename, options="rb") {

  # Check if the Rdata exist on HDFS
  hdfs_files <- hdfs_ls(filename)
  if(filename %in% names(hdfs_files)){
    message("Read file '", filename, "' on HDFS")

    # Get the file size
    file_size <- hdfs_files[[filename]]

    # Use pv to have the progress of the loading
    connection <- pipe(paste(hdfs_dfs_command(paste("-cat",filename, sep=" ")),"| pv --format '%N %b %t %r %p %e\n' --force --size",file_size, sep=" "), options)

    return(connection)
  }
  else{
    #message("File '", filename, "' not found on HDFS")
    return(FALSE)
  }
}

#' @title Write a file on HDFS
#' @name writeHDFSFile
#' @description Save a file on HDFS
#' @param filename The filename on HDFS to save
#' @param options File mode
#' @return A file object that you can use with save
writeHDFSFile <- function(filename, options="wb") {
  connection <- pipe(paste("pv --format '%N %b %t %r %p %e\n' --force |", hdfs_dfs_command("-put -"),filename, sep=" "), options)
  return(connection)
}


#' @title Make directories on HDFS
#' @name hdfs_mkdir
#' @param path Path to the directory to create
hdfs_mkdir <- function(path, debug = TRUE) {
  suppressWarnings(system(hdfs_dfs_command(paste("-mkdir", path)), ignore.stderr = debug, ignore.stdout = debug, intern = debug))
}
